cdap-app-fabric/src/main/java/co/cask/cdap/test/internal/AppFabricClient.java

```diff
@@ -69,109 +78,184 @@
+    this.programLifecycleHttpHandler = programLifecycleHttpHandler;
+    this.namespaceHttpHandler = namespaceHttpHandler;
+    this.namespaceAdmin = namespaceAdmin;
+    this.streamAdmin = streamAdmin;
+    this.streamMetaStore = streamMetaStore;
   }
 
-  public void reset() {
-    MockResponder responder = new MockResponder();
-    String uri = String.format("/v2/unrecoverable/reset");
-    HttpRequest request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.PUT, uri);
-    httpHandler.resetCDAP(request, responder);
-    verifyResponse(HttpResponseStatus.OK, responder.getStatus(), "Reset application failed");
+  public void reset() throws Exception {
+    MockResponder responder;
+    HttpRequest request;
+
+    // delete all namespaces
+    for (NamespaceMeta namespaceMeta : namespaceAdmin.listNamespaces()) {
+      Id.Namespace namespace = Id.Namespace.from(namespaceMeta.getName());
+
+      responder = new MockResponder();
+      request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.DELETE,
+                                       String.format("/v3/unrecoverable/namespaces/%s/datasets", namespace.getId()));
+      namespaceHttpHandler.deleteDatasets(request, responder, namespaceMeta.getName());
+      verifyResponse(HttpResponseStatus.OK, responder.getStatus(), String.format("could not delete datasets in namespace '%s'", namespace.getId()));
+
+      responder = new MockResponder();
+      request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.DELETE,
+                                       String.format("/v3/unrecoverable/namespaces/%s", namespace.getId()));
+      namespaceHttpHandler.delete(request, responder, namespaceMeta.getName());
+      verifyResponse(HttpResponseStatus.OK, responder.getStatus(),
+                     String.format("could not delete namespace '%s'", namespace.getId()));
+    }
   }
 
-  public void startProgram(String appId, String flowId, ProgramType type, Map<String, String> args) {
+  public void startProgram(String namespaceId, String appId,
+                           String flowId, ProgramType type, Map<String, String> args) {
 
     MockResponder responder = new MockResponder();
-    String uri = String.format("/v2/apps/%s/%s/%s/start", appId, type.getCategoryName(), flowId);
-    HttpRequest request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, uri);
-    String argString = GSON.toJson(args);
-    if (argString != null) {
-      request.setContent(ChannelBuffers.wrappedBuffer(argString.getBytes(Charsets.UTF_8)));
+    if (ProgramType.PROCEDURE.equals(type)) {
+      Preconditions.checkArgument(Constants.DEFAULT_NAMESPACE.equals(namespaceId));
+      String uri = String.format("/v2/apps/%s/%s/%s/start",
+                                 appId, type.getCategoryName(), flowId);
+      HttpRequest request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, uri);
+      String argString = GSON.toJson(args);
```

Comment:
Yeah, but not duplicating it doesn't seem like much of an improvement, and we're removing procedure support anyway, so I just did the quickest thing.